File: Assets/Script/Common/Utility/GameUtility.cs
﻿using System.IO;
using System.Linq;
using UnityEngine;

namespace Common.Utility
{
    public class GameUtility
    {
        public static byte[] SafeReadAllBytes(string inFile)
        {
            try
            {
                if (string.IsNullOrEmpty(inFile))
                {
                    return null;
                }

                if (!File.Exists(inFile))
                {
                    return null;
                }

                File.SetAttributes(inFile, FileAttributes.Normal);
                return File.ReadAllBytes(inFile);
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"SafeReadAllBytes failed! path = {inFile} with err = {ex.Message}");
                return null;
            }
        }
        /// <summary>
        /// 安全移除
        /// </summary>
        /// <param name="folderPath"></param>
        /// <returns></returns>
        public static bool LuaSafeDeleteDir(string folderPath)
        {
            try
            {
                if (string.IsNullOrEmpty(folderPath))
                {
                    return true;
                }

                if (Directory.Exists(folderPath))
                {
                    DeleteDirectory(folderPath);
                }
                return true;
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"SafeDeleteDir failed! path = {folderPath} with err: {ex.Message}");
                return false;
            }
        }

        private static void DeleteDirectory(string dirPath)
        {
            string[] files = Directory.GetFiles(dirPath);
            string[] dirs = Directory.GetDirectories(dirPath);

            foreach (string file in files)
            {
                File.SetAttributes(file, FileAttributes.Normal);
                File.Delete(file);
            }

            foreach (string dir in dirs)
            {
                DeleteDirectory(dir);
            }

            Directory.Delete(dirPath, false);
        }
        public static bool SafeClearDir(string folderPath)
        {
            try
            {
                if (string.IsNullOrEmpty(folderPath))
                {
                    return true;
                }

                if (Directory.Exists(folderPath))
                {
                    DeleteDirectory(folderPath);
                }
                Directory.CreateDirectory(folderPath);
                return true;
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"SafeClearDir failed! path = {folderPath} with err = {ex.Message}");
                return false;
            }
        }

        private static string GetFileExtension(string path)
        {
            return Path.GetExtension(path).ToLower();
        }
        public static string[] GetSpecifyFilesInFolder(string path, string[] extensions = null, bool exclude = false)
        {
            if (string.IsNullOrEmpty(path))
            {
                return null;
            }

            if (extensions == null)
            {
                return Directory.GetFiles(path, "*.*", SearchOption.AllDirectories);
            }
            else if (exclude)
            {
                return Directory.GetFiles(path, "*.*", SearchOption.AllDirectories)
                    .Where(f => !extensions.Contains(GetFileExtension(f))).ToArray();
            }
            else
            {
                return Directory.GetFiles(path, "*.*", SearchOption.AllDirectories)
                    .Where(f => extensions.Contains(GetFileExtension(f))).ToArray();
            }
        }
        public static bool LuaSafeDeleteFile(string filePath)
        {
            try
            {
                if (string.IsNullOrEmpty(filePath))
                {
                    return true;
                }

                if (!File.Exists(filePath))
                {
                    return true;
                }
                File.SetAttributes(filePath, FileAttributes.Normal);
                File.Delete(filePath);
                return true;
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"SafeDeleteFile failed! path = {filePath} with err: {ex.Message}");
                return false;
            }
        }
        
        public static bool LuaRenameFile(string sourceFileName, string destFileName)
        {
            try
            {
                if (string.IsNullOrEmpty(sourceFileName))
                {
                    return false;
                }

                if (!File.Exists(sourceFileName))
                {
                    return true;
                }
                LuaSafeDeleteFile(destFileName);
                File.SetAttributes(sourceFileName, FileAttributes.Normal);
                File.Move(sourceFileName, destFileName);
                return true;
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"SafeRenameFile failed! path = {sourceFileName} with err: {ex.Message}");
                return false;
            }
        }
    }
    
}

File: Assets/.lpproj/Apis/Unity/CS/Common/Utility/GameUtility.lua
---@class CS.Common.Utility.GameUtility : CS.System.Object
CS.Common.Utility.GameUtility = {}

---@field public CS.Common.Utility.GameUtility.AssetsFolderName : CS.System.String
CS.Common.Utility.GameUtility.AssetsFolderName = nil

---@return CS.Common.Utility.GameUtility
function CS.Common.Utility.GameUtility()
end

---@param inFile : CS.System.String
---@return CS.System.Byte[]
function CS.Common.Utility.GameUtility.SafeReadAllBytes(inFile)
end

---@param folderPath : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.LuaSafeDeleteDir(folderPath)
end

---@param folderPath : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeClearDir(folderPath)
end

---@param path : CS.System.String
---@return CS.System.String
function CS.Common.Utility.GameUtility.FormatToSysFilePath(path)
end

---@param path : CS.System.String
---@param extensions : CS.System.String[]
---@param exclude : CS.System.Boolean
---@return CS.System.String[]
function CS.Common.Utility.GameUtility.GetSpecifyFilesInFolder(path, extensions, exclude)
end

---@param filePath : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.LuaSafeDeleteFile(filePath)
end

---@param sourceFileName : CS.System.String
---@param destFileName : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.LuaRenameFile(sourceFileName, destFileName)
end

---@param outFile : CS.System.String
---@param outBytes : CS.System.Byte[]
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeWriteAllBytes(outFile, outBytes)
end

---@param inFile : CS.System.String
---@return CS.System.String
function CS.Common.Utility.GameUtility.SafeReadAllText(inFile)
end

---@param folderPath : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeDeleteDir(folderPath)
end

---@param outFile : CS.System.String
---@param text : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeWriteAllText(outFile, text)
end

---@param outFile : CS.System.String
---@param outLines : CS.System.String[]
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeWriteAllLines(outFile, outLines)
end

---@param path : CS.System.String
---@return CS.System.String
function CS.Common.Utility.GameUtility.FormatToUnityPath(path)
end

---@param full_path : CS.System.String
---@return CS.System.String
function CS.Common.Utility.GameUtility.FullPathToAssetPath(full_path)
end

---@param folderPath : CS.System.String
function CS.Common.Utility.GameUtility.CheckDirAndCreateWhenNeeded(folderPath)
end

---@param filePath : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeDeleteFile(filePath)
end

---@param sourceFileName : CS.System.String
---@param destFileName : CS.System.String
---@return CS.System.Boolean
function CS.Common.Utility.GameUtility.SafeRenameFile(sourceFileName, destFileName)
end